bpo-45530: speed listobject.c's unsafe_tuple_compare()

tim-one opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->
[bpo-45530](https://bugs.python.org/issue45530): speed listobject.c's unsafe_tuple_compare()


<!-- issue-number: [bpo-45530](https://bugs.python.org/issue45530) -->
https://bugs.python.org/issue45530
<!-- /issue-number -->

tim-one committed:
First stab. About 40% speedup on tupsort.py's "(float,)" case.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

rhettinger commented on Objects/listobject.c:
```diff
@@ -2195,25 +2195,32 @@ unsafe_tuple_compare(PyObject *v, PyObject *w, MergeState *ms)
             return -1;
-        if (!k)
-            break;
+        if (!k) {
```
I would use "k > 0" here for clarity (and for consistency with the code above).

tim-one replied:
But that doesn't make sense: `!k` is the same as `k == 0`, not `k > 0`. `k == 0` in turn means "`Py_EQ` is not the case - they're not equal". And I never write `k == 0` in C, because it's too frickin' likely I'll type `k = 0` instead :wink:.

ambv committed:
Add reordering notification to whatsnew, clarify Blurb a little

ambv committed:
Fix invalid ref in Blurb

tim-one committed:
Keep track of whether unsafe_tuple_compare() calls are
resolved by the very first tuple elements, and adjust
strategy accordingly.

tim-one committed:
Add commenra explaining the ms->first_tuple_items_resolved_it
strategy. This looks to be quite successful. It loses a few
per cent in speed in cases that always want to use the cheaper
tests, but can gain far more (compared to this branch's state
before this commit) in some cases where
PyObject_RichCompareBool(..., Py_EQ) typically returns 1 (they're
equal) when applied to the first pair.

tim-one committed:
Add a clariying assert.

Swap the order of if/else blocks to put the more likely
block first.

tim-one committed:
Remove the new `firsti` vrbl - needless name proliferation.

tim-one committed:
Errors from comparisons are rare, so rearrange code to act on normal outcomes first.
